pkg/filters/grpcproxy/pool.go

```diff
@@ -0,0 +1,485 @@
+		logger.Debugf("%s: short circuited by circuit break policy", sp.name)
+		spCtx.AddTag("short circuited")
+		sp.buildOutputResponse(spCtx, status.Newf(codes.Unavailable, "short circuited by circuit break policy"))
+		return resultShortCircuited
+	}
+
+	// The error must be a serverPoolError now, we need to build a
+	// response in most cases, but for failure status codes, the
+	// response is already there.
+	if spe, ok := err.(serverPoolError); ok {
+		sp.buildOutputResponse(spCtx, spe.status)
+		return spe.Result()
+	}
+
+	panic(fmt.Errorf("should not reach here"))
+}
+
+func (sp *ServerPool) doHandle(ctx stdcontext.Context, spCtx *serverPoolContext) error {
+	lb := sp.LoadBalancer()
+	svr := lb.ChooseServer(spCtx.req)
+	// if there's no available server.
+	if svr == nil {
+		logger.Debugf("%s: no available server", sp.name)
+		return serverPoolError{status.New(codes.InvalidArgument, "no available server"), resultClientError}
+	}
+	if f, ok := lb.(ReusableServerLB); ok {
+		defer f.ReturnServer(svr)
+	}
+	// maybe be rewrite by grpcserver.MuxPath#rewrite
+	fullMethodName := spCtx.req.FullMethod()
+	if fullMethodName == "" {
+		return serverPoolError{status.New(codes.InvalidArgument, "unknown called method from context"), resultClientError}
+	}
+	send2ProviderCtx, cancelContext := stdcontext.WithCancel(metadata.NewOutgoingContext(ctx, spCtx.req.RawHeader().GetMD()))
+	defer cancelContext()
+	dialCtx, cancel := stdcontext.WithCancel(stdcontext.Background())
+	if sp.spec.BlockUntilConnected {
+		dialCtx, cancel = stdcontext.WithTimeout(dialCtx, sp.connectTimeout)
+	}
+	defer cancel()
+	conn, err := grpc.DialContext(dialCtx, svr.URL, sp.dialOpts...)
+	if err != nil {
+		logger.Infof("create new conn without pool fail %s for source addr %s, target addr %s, path %s",
+			err.Error(), spCtx.req.SourceHost(), svr.URL, fullMethodName)
+		return serverPoolError{status: status.Convert(err), result: resultInternalError}
+	}
+	proxyAsClientStream, err := conn.NewStream(send2ProviderCtx, desc, fullMethodName)
+	defer conn.Close()
+	if err != nil {
+		logger.Infof("create new stream fail %s for source addr %s, target addr %s, path %s",
+			err.Error(), spCtx.req.SourceHost(), svr.URL, fullMethodName)
+		return serverPoolError{status: status.Convert(err), result: resultInternalError}
+	}
+
+	result := sp.biTransport(spCtx, proxyAsClientStream)
+	if result != nil && result != io.EOF {
+		logger.Infof("create new stream fail %s for source addr %s, target addr %s, path %s",
+			result.Error(), spCtx.req.SourceHost(), svr.URL, fullMethodName)
+	}
+	return result
+}
+
+func (sp *ServerPool) biTransport(ctx *serverPoolContext, proxyAsClientStream grpc.ClientStream) error {
+	// Explicitly *do not Close* c2sErrChan and c2sErrChan, otherwise the select below will not terminate.
+	// Channels do not have to be closed, it is just a control flow mechanism, see
+	// https://groups.google.com/forum/#!msg/golang-nuts/pZwdYRGxCIk/qpbHxRRPJdUJ
```

Comment:
it is better to close the channels in the goroutine created by `sp.forwardE2E`, just before it exit.

Reply:
closing the channel is not always a good option in this scenario. because the client indicates that it will not continue to send datato server on the ClientStream after called `ClientStream.CloseSend`, the server can take a long time to handle business before return IO.EOF or continue to send data in one direction, and the client can also continue processing , so C2S and S2C are not always sync. In this scenario, `select in for loop` will always select the closed channel and take a lot of CPU